File: frontend/src/types/visualization/sentinel2.ts
import type {ButtonOption} from "../../components/sidebar/ButtonGroup.tsx";


// ======================================================
// BANDS
// ======================================================

export interface Sentinel2BandDefinition {
    id: string;
    label: string;
}


export const SENTINEL2_BANDS = [
    {
        id: "1",
        label: "B1 - Coastal Aerosol",
    },
    {
        id: "2",
        label: "B2 - Blue",
    },
    {
        id: "3",
        label: "B3 - Green",
    },
    {
        id: "4",
        label: "B4 - Red",
    },
    {
        id: "5",
        label: "B5 - Vegetation Red Edge",
    },
    {
        id: "6",
        label: "B6 - Vegetation Red Edge",
    },
    {
        id: "7",
        label: "B7 - Vegetation Red Edge",
    },
    {
        id: "8",
        label: "B8 - NIR",
    },
    {
        id: "8A",
        label: "B8A - Narrow NIR",
    },
    {
        id: "9",
        label: "B9 - Water Vapour",
    },
    {
        id: "10",
        label: "B10 - Cirrus",
    },
    {
        id: "11",
        label: "B11 - SWIR 1",
    },
    {
        id: "12",
        label: "B12 - SWIR 2",
    },
    {
        id: "TCI",
        label: "TCI - True Color Image",
    },
] as const satisfies readonly Sentinel2BandDefinition[];


export type Sentinel2Band = typeof SENTINEL2_BANDS[number]["id"];


// UI adapter

export const SENTINEL2_BAND_OPTIONS = SENTINEL2_BANDS.map(
    band => ({
        value: band.id,
        label: band.label,
    }),
) satisfies readonly ButtonOption<Sentinel2Band>[];


// ======================================================
// SPECTRAL BANDS
// ======================================================

export const SENTINEL2_SPECTRAL_BANDS = SENTINEL2_BANDS.filter(
    band => band.id !== "TCI",
);


export type Sentinel2SpectralBand = typeof SENTINEL2_SPECTRAL_BANDS[number]["id"];


export const SENTINEL2_SPECTRAL_BAND_OPTIONS = SENTINEL2_SPECTRAL_BANDS.map(
    band => ({
        value: band.id,
        label: band.label,
    }),
) satisfies readonly ButtonOption<Sentinel2SpectralBand>[];


// ======================================================
// VISUALIZATION MODE
// ======================================================

export const SENTINEL2_VISUALIZATION_MODE = {
    SINGLE_BANDS: "single-bands",
    CUSTOM_RGB: "custom-rgb",
    PRESETS: "presets",
} as const;


export type Sentinel2VisualizationMode = typeof SENTINEL2_VISUALIZATION_MODE[keyof typeof SENTINEL2_VISUALIZATION_MODE];


// ======================================================
// RGB COMPOSITE
// ======================================================

export interface Sentinel2RGBComposite {
    red: Sentinel2SpectralBand;
    green: Sentinel2SpectralBand;
    blue: Sentinel2SpectralBand;
}


export const DEFAULT_SENTINEL2_RGB_COMPOSITE: Sentinel2RGBComposite = {
    red: "4",
    green: "3",
    blue: "2",
};


// ======================================================
// FEATURES / SPECTRAL INDICES
// ======================================================

export const SENTINEL2_INDEX = {
    HONS: "HONS",
    NDVI: "NDVI",
    NDWI: "NDWI",
    NDMI: "NDMI",
    NDSI: "NDSI",
    ND_UNSPEC: "ND_UNSPEC",
} as const;


export type Sentinel2Index = typeof SENTINEL2_INDEX[keyof typeof SENTINEL2_INDEX];


// ======================================================
// REQUIRED BANDS FOR FIXED FEATURES
// ======================================================

/**
 * Required input bands for features with a fixed definition.
 *
 * The order is important because it corresponds to the order
 * expected by gjtiff:
 *
 * HONS      -> B04, B03, B02
 * NDVI      -> B08, B04
 * NDWI      -> B03, B08
 * NDMI      -> B8A, B11
 * NDSI      -> B03, B11, B04, B02
 */
export const SENTINEL2_INDEX_BANDS = {
    [SENTINEL2_INDEX.HONS]: ["4", "3", "2"],
    [SENTINEL2_INDEX.NDVI]: ["8", "4"],
    [SENTINEL2_INDEX.NDWI]: ["3", "8"],
    [SENTINEL2_INDEX.NDMI]: ["8A", "11"],
    [SENTINEL2_INDEX.NDSI]: ["3", "11", "4", "2"],
} as const satisfies Record<
    Exclude<Sentinel2Index, typeof SENTINEL2_INDEX.ND_UNSPEC>,
    readonly Sentinel2SpectralBand[]
>;


// ======================================================
// CUSTOM NORMALIZED DIFFERENCE
// ======================================================

/**
 * Generic normalized difference:
 *
 *     (Bx - By) / (Bx + By)
 *
 * Unlike the fixed spectral indices, the user chooses
 * both input bands.
 */
export interface Sentinel2CustomNormalizedDifference {
    band1: Sentinel2SpectralBand;
    band2: Sentinel2SpectralBand;
}


// ======================================================
// INDEX VISUALIZATION
// ======================================================

export interface Sentinel2IndexVisualization {
    index: Sentinel2Index;

    /**
     * Only used for ND_UNSPEC.
     */
    band1?: Sentinel2SpectralBand;
    band2?: Sentinel2SpectralBand;
}


// ======================================================
// PRESETS
// ======================================================

export const SENTINEL2_PRESET_TYPE = {
    RGB_COMPOSITE: "rgb-composite",
    INDEX: "index",
} as const;


export type Sentinel2PresetType = typeof SENTINEL2_PRESET_TYPE[keyof typeof SENTINEL2_PRESET_TYPE];


export interface Sentinel2PresetBase {
    id: string;
    label: string;
    presetType: Sentinel2PresetType;
}


export interface Sentinel2RGBPreset extends Sentinel2PresetBase {
    presetType: typeof SENTINEL2_PRESET_TYPE.RGB_COMPOSITE;
    composite: Sentinel2RGBComposite;
}


export interface Sentinel2IndexPreset extends Sentinel2PresetBase {
    presetType: typeof SENTINEL2_PRESET_TYPE.INDEX;
    index: Sentinel2Index;
}


export type Sentinel2Preset = Sentinel2RGBPreset | Sentinel2IndexPreset;


// ======================================================
// PRESET DEFINITIONS
// ======================================================

export const SENTINEL2_PRESETS = [
    // --------------------------------------------------
    // RGB composites
    // --------------------------------------------------

    {
        id: "true-color",
        label: "True Color",

        presetType: SENTINEL2_PRESET_TYPE.RGB_COMPOSITE,

        composite: {
            red: "4",
            green: "3",
            blue: "2",
        },
    },

    {
        id: "false-color-infrared",
        label: "False Color Infrared",

        presetType: SENTINEL2_PRESET_TYPE.RGB_COMPOSITE,

        composite: {
            red: "8",
            green: "4",
            blue: "3",
        },
    },

    {
        id: "agriculture",
        label: "Agriculture",

        presetType: SENTINEL2_PRESET_TYPE.RGB_COMPOSITE,

        composite: {
            red: "11",
            green: "8",
            blue: "2",
        },
    },

    {
        id: "geology",
        label: "Geology",

        presetType: SENTINEL2_PRESET_TYPE.RGB_COMPOSITE,

        composite: {
            red: "12",
            green: "11",
            blue: "2",
        },
    },

    // --------------------------------------------------
    // Spectral / combined features
    // --------------------------------------------------

    {
        id: "hons",
        label: "HONS",

        presetType: SENTINEL2_PRESET_TYPE.INDEX,

        index: SENTINEL2_INDEX.HONS,
    },

    {
        id: "ndvi",
        label: "NDVI",

        presetType: SENTINEL2_PRESET_TYPE.INDEX,

        index: SENTINEL2_INDEX.NDVI,
    },

    {
        id: "ndwi",
        label: "NDWI",

        presetType: SENTINEL2_PRESET_TYPE.INDEX,

        index: SENTINEL2_INDEX.NDWI,
    },

    {
        id: "ndmi",
        label: "NDMI",

        presetType: SENTINEL2_PRESET_TYPE.INDEX,

        index: SENTINEL2_INDEX.NDMI,
    },

    {
        id: "ndsi",
        label: "NDSI",

        presetType: SENTINEL2_PRESET_TYPE.INDEX,

        index: SENTINEL2_INDEX.NDSI,
    },
] as const satisfies readonly Sentinel2Preset[];


export type Sentinel2PresetId = typeof SENTINEL2_PRESETS[number]["id"];


// UI adapter

export const SENTINEL2_PRESET_OPTIONS =
    SENTINEL2_PRESETS.map(
        preset => ({
            value: preset.id,
            label: preset.label,
        }),
    ) satisfies readonly ButtonOption<Sentinel2PresetId>[];


// ======================================================
// BACKEND PAYLOAD TYPES
// ======================================================

export interface Sentinel2RGBVisualization {
    red: Sentinel2SpectralBand;
    green: Sentinel2SpectralBand;
    blue: Sentinel2SpectralBand;
}


export interface Sentinel2Visualizations {
    /**
     * Individual bands to generate.
     *
     * Example:
     * ["4", "8", "11"]
     */
    bands?: Sentinel2SpectralBand[];

    /**
     * Custom RGB composite.
     *
     * Example:
     * {
     *     red: "8",
     *     green: "4",
     *     blue: "3"
     * }
     */
    rgb?: Sentinel2RGBVisualization;

    /**
     * Spectral / combined features.
     *
     * Fixed features:
     *     { feature: "NDVI" }
     *     { feature: "NDWI" }
     *     { feature: "NDMI" }
     *     { feature: "NDSI" }
     *     { feature: "HONS" }
     *
     * Custom:
     *     {
     *         feature: "ND_UNSPEC",
     *         band1: "8",
     *         band2: "4"
     *     }
     */
    features?: Sentinel2IndexVisualization[];

    /**
     * Named presets.
     */
    presets?: Sentinel2PresetId[];
}

File: backend/app/domain/dataset/sentinel_2.py
from dataclasses import dataclass
from enum import StrEnum


# ======================================================
# BANDS
# ======================================================

class Sentinel2Band(StrEnum):
    B1 = "B01"
    B2 = "B02"
    B3 = "B03"
    B4 = "B04"
    B5 = "B05"
    B6 = "B06"
    B7 = "B07"
    B8 = "B08"
    B8A = "B8A"
    B9 = "B09"
    B10 = "B10"
    B11 = "B11"
    B12 = "B12"
    TCI = "TCI"


# ======================================================
# RGB
# ======================================================

@dataclass(frozen=True, slots=True)
class Sentinel2RGBComposite:
    red: Sentinel2Band
    green: Sentinel2Band
    blue: Sentinel2Band


# ======================================================
# INDEXES
# ======================================================

class Sentinel2Index(StrEnum):
    HONS = "HONS"
    NDVI = "NDVI"
    NDWI = "NDWI"
    NDMI = "NDMI"
    NDSI = "NDSI"
    ND_UNSPEC = "ND_UNSPEC"


# Band order is aligned with GJTIFF input.
SENTINEL2_INDEX_BANDS: dict[Sentinel2Index, tuple[Sentinel2Band, ...]] = {
    Sentinel2Index.HONS: (
        Sentinel2Band.B4,
        Sentinel2Band.B3,
        Sentinel2Band.B2,
    ),

    Sentinel2Index.NDVI: (
        Sentinel2Band.B8,
        Sentinel2Band.B4,
    ),

    Sentinel2Index.NDWI: (
        Sentinel2Band.B3,
        Sentinel2Band.B8,
    ),

    Sentinel2Index.NDMI: (
        Sentinel2Band.B8A,
        Sentinel2Band.B11,
    ),

    Sentinel2Index.NDSI: (
        Sentinel2Band.B3,
        Sentinel2Band.B11,
        Sentinel2Band.B4,
        Sentinel2Band.B2,
    ),
}


# ======================================================
# PRESETS
# ======================================================

class Sentinel2PresetType(StrEnum):
    RGB_COMPOSITE = "rgb-composite"
    INDEX = "index"


@dataclass(frozen=True, slots=True)
class Sentinel2PresetBase:
    id: str
    label: str
    preset_type: Sentinel2PresetType


@dataclass(frozen=True, slots=True)
class Sentinel2RGBPreset(Sentinel2PresetBase):
    composite: Sentinel2RGBComposite


@dataclass(frozen=True, slots=True)
class Sentinel2IndexPreset(Sentinel2PresetBase):
    index: Sentinel2Index


SENTINEL2_PRESETS = {
    # ======================================================
    # RGB PRESETS
    # ======================================================

    "true-color": Sentinel2RGBPreset(
        id="true-color",
        label="True Color",
        preset_type=Sentinel2PresetType.RGB_COMPOSITE,
        composite=Sentinel2RGBComposite(
            red=Sentinel2Band.B4,
            green=Sentinel2Band.B3,
            blue=Sentinel2Band.B2,
        ),
    ),

    "false-color-infrared": Sentinel2RGBPreset(
        id="false-color-infrared",
        label="False Color Infrared",
        preset_type=Sentinel2PresetType.RGB_COMPOSITE,
        composite=Sentinel2RGBComposite(
            red=Sentinel2Band.B8,
            green=Sentinel2Band.B4,
            blue=Sentinel2Band.B3,
        ),
    ),

    "agriculture": Sentinel2RGBPreset(
        id="agriculture",
        label="Agriculture",
        preset_type=Sentinel2PresetType.RGB_COMPOSITE,
        composite=Sentinel2RGBComposite(
            red=Sentinel2Band.B11,
            green=Sentinel2Band.B8,
            blue=Sentinel2Band.B2,
        ),
    ),

    "geology": Sentinel2RGBPreset(
        id="geology",
        label="Geology",
        preset_type=Sentinel2PresetType.RGB_COMPOSITE,
        composite=Sentinel2RGBComposite(
            red=Sentinel2Band.B12,
            green=Sentinel2Band.B11,
            blue=Sentinel2Band.B2,
        ),
    ),

    # ======================================================
    # INDEX PRESETS
    # ======================================================

    "hons": Sentinel2IndexPreset(
        id="hons",
        label="HONS",
        preset_type=Sentinel2PresetType.INDEX,
        index=Sentinel2Index.HONS,
    ),

    "ndvi": Sentinel2IndexPreset(
        id="ndvi",
        label="NDVI",
        preset_type=Sentinel2PresetType.INDEX,
        index=Sentinel2Index.NDVI,
    ),

    "ndwi": Sentinel2IndexPreset(
        id="ndwi",
        label="NDWI",
        preset_type=Sentinel2PresetType.INDEX,
        index=Sentinel2Index.NDWI,
    ),

    "ndmi": Sentinel2IndexPreset(
        id="ndmi",
        label="NDMI",
        preset_type=Sentinel2PresetType.INDEX,
        index=Sentinel2Index.NDMI,
    ),

    "ndsi": Sentinel2IndexPreset(
        id="ndsi",
        label="NDSI",
        preset_type=Sentinel2PresetType.INDEX,
        index=Sentinel2Index.NDSI,
    ),
}


# ======================================================
# DOMAIN HELPERS
# ======================================================

def get_required_sentinel2_bands(
        visualizations: dict,
) -> set[Sentinel2Band]:
    """
    Return all Sentinel-2 bands required by a visualization request.
    """

    required_bands: set[Sentinel2Band] = set()

    # Single bands
    for band in visualizations.get("bands", []):
        try:
            required_bands.add(Sentinel2Band(band))
        except ValueError:
            continue

    # Custom RGB composite
    rgb = visualizations.get("rgb_composite")

    if rgb:
        for channel in ("red", "green", "blue"):
            try:
                required_bands.add(Sentinel2Band(rgb[channel]))

            except (KeyError, ValueError):
                continue

    # Presets
    for preset_id in visualizations.get("presets", []):
        preset = SENTINEL2_PRESETS.get(preset_id)

        if preset is None:
            continue

        if isinstance(preset, Sentinel2RGBPreset):
            required_bands.update({
                preset.composite.red,
                preset.composite.green,
                preset.composite.blue,
            })

        elif isinstance(preset, Sentinel2IndexPreset):
            required_bands.update(
                SENTINEL2_INDEX_BANDS[preset.index]
            )

    return required_bands
